enhancements/support-for-realtime-kernel.md

```diff
@@ -0,0 +1,292 @@
+
+## Design Details
+
+### Providing the Packages
+
+The proposal is to include the `kernel-rt` packages in the `machine-os-content`
+image.  They **WILL NOT** be included as part of the ostree commit, but rather
+as separate files within the image.  Example:
+
+```
+$ sudo podman pull quay.io/openshift-release-dev/ocp-v4.0-art-dev:realtime
+Trying to pull quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:realtime...
+Getting image source signatures
+Copying blob 3c141bccb4e4 skipped: already exists
+Copying config 485b845970 done
+Writing manifest to image destination
+Storing signatures
+485b845970f19098b662f85a9e22bf4db7aabea83fe5d9c155d798dffcd6f2e9
+
+$ ctr=$(sudo podman create quay.io/openshift-release-dev/ocp-v4.0-art-dev:realtime)
+
+$ mnt=$(sudo podman mount $ctr)
+
+$ sudo ls -l $mnt
+total 50036
+-rw-r--r--. 1 miabbott miabbott 26061020 Dec 16 15:58 kernel-rt-core-4.18.0-147.0.3.rt24.95.el8_1.x86_64.rpm
+-rw-r--r--. 1 miabbott miabbott 22869480 Dec 16 15:58 kernel-rt-modules-4.18.0-147.0.3.rt24.95.el8_1.x86_64.rpm
+-rw-r--r--. 1 miabbott miabbott  2285240 Dec 16 15:58 kernel-rt-modules-extra-4.18.0-147.0.3.rt24.95.el8_1.x86_64.rpm
+-rw-rw-r--. 1 root     root        13764 Dec 16 15:59 pkglist.txt
+drwxrwxr-x. 3 root     root           18 Dec 16 15:58 srv
+
+```
+
+### Selecting the Kernel
+
+The proposal is to have the MachineConfig spec changed to support a field named
+`kernelType` which will determine the which kernel should be used on the RHCOS
+nodes. If `kernelType: realtime` is configured, the `kernel-rt` packages will
+be used.  If `kernelType: default` is configured, the default kernel will be used.
+In the absence of any `kernelType` field, the default choice is for the default
+kernel to be used.
+
+Example MachineConfig:
+
+```
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: "worker"
+  name: 99-worker-realtime-kernel
+spec:
+  kernelType: "realtime"
+```
+
+### Installing the Real-time Kernel
+
+When the MCO parses a MachineConfig with `kernelType: realtime`, it shall instruct
+`rpm-ostree` on the RHCOS node to override the installed kernel with the `kernel-rt`
+package from the `machine-os-content` image.
+
+### Removing the Real-time Kernel
+
+When the MCO parses a MachineConfig with `kernelType: default`, it shall instruct `rpm-ostree`
+on the RHCOS node to remove any `kernel-rt` packages and use the default kernel.  If the
+`kernel-rt` packages are not present, it should be a no-op.
+
+### Upgrades
+
+When the OCP cluster is upgraded, the RHCOS nodes that are using the real-time
+kernel will also update the `kernel-rt` packages if updates are available in
+the `machine-os-content` image.  This treats the `kernel-rt` package like any
+other RPM installed on the RHCOS nodes.
+
+### Test Plan
+
+- Test install of `kernel-rt` packages on single RHCOS node
+- Test removal of `kernel-rt` packages on single RHCOS node
```

Comment:
Test upgrade of `kernel-rt` package on a single RHCOS node?  So we can possible catch breakage earlier than when it gets into OCP.  We'd need another version but same issue applies with updating OCP with a newer `kernel-rt`